src/registers/mxcsr.rs

```diff
@@ -0,0 +1,107 @@
+//! Functions to read and write MXCSR register.
+
+#[cfg(feature = "instructions")]
+pub use self::x86_64::*;
+
+use bitflags::bitflags;
+
+bitflags! {
+    /// MXCSR register.
+    #[repr(transparent)]
+    pub struct MxCsr: u32 {
+        /// Invalid operation flag
+        const INVALID_OPERATION = 1 << 0;
+        /// Denormal flag
+        const DENORMAL = 1 << 1;
+        /// Divide-by-zero flag
+        const DIVIDE_BY_ZERO = 1 << 2;
+        /// Overflow flag
+        const OVERFLOW = 1 << 3;
+        /// Underflow flag
+        const UNDERFLOW = 1 << 4;
+        /// Precision flag
+        const PRECISION = 1 << 5;
+        /// Denormals are zeros
+        const DENORMALS_ARE_ZEROS = 1 << 6;
+        /// Invalid operation mask
+        const INVALID_OPERATION_MASK = 1 << 7;
+        /// Denormal mask
+        const DENORMAL_MASK = 1 << 8;
+        /// Divide-by-zero mask
+        const DIVIDE_BY_ZERO_MASK = 1 << 9;
+        /// Overflow mask
+        const OVERFLOW_MASK = 1 << 10;
+        /// Underflow mask
+        const UNDERFLOW_MASK = 1 << 11;
+        /// Precision mask
+        const PRECISION_MASK = 1 << 12;
+        /// Toward negative infinity
+        const ROUNDING_CONTROL_NEGATIVE = 1 << 13;
+        /// Toward positive infinity
+        ///
+        /// Set the both control registers for "toward zero".
+        const ROUNDING_CONTROL_POSITIVE = 1 << 14;
+        /// Flush to zero
+        const FLUSH_TO_ZERO = 1 << 15;
+    }
+}
+
+impl Default for MxCsr {
+    /// Return the default MXCSR value at reset, as documented in Intel SDM volume 2A.
+    #[inline]
+    fn default() -> Self {
+        MxCsr::from_bits_truncate(0x1F80)
+    }
+}
+
+#[cfg(feature = "instructions")]
+mod x86_64 {
+    use super::*;
+    #[cfg(feature = "inline_asm")]
+    use core::arch::asm;
+
+    /// Read the value of MXCSR.
+    #[inline]
+    pub fn read() -> MxCsr {
+        let mut mxcsr: u32 = 0;
+
+        unsafe {
+            asm!("stmxcsr [{}]", in(reg) &mut mxcsr, options(nostack));
+        }
+
+        MxCsr::from_bits_truncate(mxcsr)
+    }
+
+    /// Write MXCSR.
+    ///
+    /// # Safety
+    ///
+    /// Unsafe because undefined becavior can occur given that input is not
+    /// validate before it is passed to ldmxcsr.
```

Comment:
How can this cause undefined behavior (eg which flags)?

Reply:
The upper 16-bit word contains reserved bits (cleared on power on), and setting them causes #GP. Should pre-validate or just document?

Reply:
We don't consider triggering CPU exceptions to be unsafe (see https://github.com/rust-osdev/x86_64/issues/202#issuecomment-719923562). Are there any other ways this could cause UB?

Reply:
Nothing else comes to mind, e.g. according to the SDM:

```
Flags Affected
None
```

Reply:
Right, setting upper flags would be _well-defined_ behavior because you consistently SEGFAULT :-) I got it.